pkg/upgrade/machine_creator.go

```diff
@@ -0,0 +1,283 @@
+
+	if n.MachineOptions.ImageField != "" && n.MachineOptions.ImageID != "" {
+		if err := updateMachineSpecImage(&newMachine.Spec, n.MachineOptions.ImageField, n.MachineOptions.ImageID); err != nil {
+			return nil, nil, err
+		}
+	}
+
+	newMachine.Spec.Versions.ControlPlane = n.MachineOptions.DesiredVersion.String()
+	newMachine.Spec.Versions.Kubelet = n.MachineOptions.DesiredVersion.String()
+
+	n.log.Info("Creating new machine", "name", newMachine.GetName())
+
+	createdMachine, err := n.machineCreator.Create(newMachine)
+	if err != nil {
+		return nil, nil, errors.Wrapf(err, "Error creating machine: %s", newMachine.Name)
+	}
+
+	if n.WaitForProviderID {
+		providerID, err := n.waitForProviderID(createdMachine.GetNamespace(), createdMachine.GetName(), 15*time.Minute)
+		if err != nil {
+			return nil, nil, err
+		}
+		if n.WaitForMatchingNode {
+			node, err := n.waitForMatchingNode(providerID, 5*time.Minute)
+			if err != nil {
+				return nil, nil, err
+			}
+			if n.WaitForNodeReady {
+				if err := n.waitForNodeReady(node); err != nil {
+					return nil, nil, err
+				}
+				// return ready node (could delete but here for explicitness)
+				return createdMachine, node, nil
+			}
+			// return an unready node
+			return createdMachine, node, nil
+		}
+		// return the created machine with provider ID and no node since we are not waiting for the node
+		return createdMachine, nil, nil
+	}
+	// return the created machine without a provider ID and no node since we waited for nothing
+	return createdMachine, nil, nil
+}
+
+func (n *MachineCreator) waitForProviderID(ns, name string, timeout time.Duration) (string, error) {
+	n.log.Info("waitForMachineProviderID start", "namespace", ns, "name", name)
+	var providerID string
+	err := wait.PollImmediate(5*time.Second, timeout, func() (bool, error) {
+		machine, err := n.machineGetter.Get(name, metav1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
+
+		if machine.Spec.ProviderID == nil {
+			return false, nil
+		}
+
+		providerID = *machine.Spec.ProviderID
+		if providerID != "" {
+			n.log.Info("Got providerID", "provider-id", providerID)
+			return true, nil
+		}
+		return false, nil
+	})
+
+	if err != nil {
+		return "", errors.Wrap(err, "timed out waiting for machine provider id")
+	}
+
+	p, err := kubernetes.ParseProviderID(providerID)
+	if err != nil {
+		n.log.Error(err, "error parsing new provider id", "provider-id", providerID)
+	}
+	return p, nil
+}
+
+func (n *MachineCreator) waitForMatchingNode(providerID string, timeout time.Duration) (*v1.Node, error) {
+	n.log.Info("Waiting for node", "provider-id", providerID)
+	var matchingNode v1.Node
+	err := wait.PollImmediate(5*time.Second, timeout, func() (bool, error) {
+		nodes, err := n.nodeLister.List(metav1.ListOptions{})
+		if err != nil {
+			return false, err
+		}
+		for _, node := range nodes.Items {
+			if node.Spec.ProviderID == providerID {
```

Comment:
Now that the provider id code is in the CAPI `release-0.1` branch, we'll want to switch to using it. This line here won't work for e.g. AWS because the node's provider id will be something like `aws://us-east-1/i-xxxxx` and the machine's provider id doesn't contain the region.

Reply:
oh ok, could you open an issue with the details? Seems out of scope for this PR

Reply:
Sure, but this is a regression if you leave it like this. For now, you'll need to run `kubernetes.ParseProviderID(node.Spec.ProviderID)` and compare it with `providerID`. Make sense?

Reply:
oh, hmm, we are doing that, the parsed provider ID comes back from `waitForProviderID` https://github.com/vmware/cluster-api-upgrade-tool/pull/47/files#diff-d11660ea136b9301cfeac8702fc73ca1R157

Reply:
You have to parse `node.Spec.ProviderID` here, which is missing

Reply:
Ah. yes. totally. my fault, excellent code reading skills

Reply:
Instead of parsing *both* I think it makes more sense to parse neither and do less work.

since `providerID` is already parsed and I have to parse `node.Spec.ProviderID`, this could instead parse neither  -- or is there something i'm missing from ProviderIDs?

Reply:
The k/k cloud provider, at least for AWS, includes the region in the provider id. CAPA does not include the region. A straight up string comparison will not work. The best solution is to use the `Equals()` function from https://github.com/kubernetes-sigs/cluster-api/blob/master/pkg/controller/noderefutil/providerid.go.

Reply:
ok, thanks for explaining~! got it